### PR TITLE
Issue #949 allocate river layers

### LINKED ISSUE #949: Create utility functions to allocate RIV/DRN/GHB cells
See this report from Sweco describing different methods to allocate cells with depth and distribute conductances over depth in iMOD5. Nice summary of requirements for this story!

[20240129 Conductanceverdeling iMOD5.docx](https://github.com/Deltares/imod-python/files/14222631/20240129.Conductanceverdeling.iMOD5.docx)

I translated the list of settings in iMOD5.6 briefly below. We could create a set of utility functions to support this, so that users can mix and match what they'd like.

Allocation RIV cells with depth:

* Laying inbetween river stage and bottom elevation (IDEFFLAYER=0)
* First active cell to bottom elevation (IDEFFLAYER=-1)
* First active cell to bottom elevation, cells above river stage can only drain (IDEFFLAYER=1)
* In cells where bottom elevation is located (IDEFFLAYER=2)

### PR DESCRIPTION
**Pull request by JoerivanEngelen**
Fixes #949 

# Description

- Adds functionality to allocate river cells. 
- These can also be used to allocate DRN, GHB, and RCH cells
- I had to move the ``polygonize`` util function to work around a nasty circular import. The solution also prevents changes to the public API.
- A decorator was added to enforce an iMOD Python compatible dimension order (Related to: #239 )
- A decorator was added to preserve grid type (xu.UgridDataArray or xr.DataArray).

# Checklist

- [x] Links to correct issue
- [x] Update changelog, if changes affect users
- [x] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [x] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/prepare/topsystem/allocation.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+    Parameters
+    ----------
+    active: GridDataArray
+        active model cells
+    top: GridDataArray
+        top of model layers
+    bottom: GridDataArray
+        bottom of model layers
+    bottom_elevation: GridDataArray
```
Comment (luitjansl) [inline imod/prepare/topsystem/allocation.py:192]: river_bottom_elevation?

File: imod/prepare/topsystem/allocation.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+            )
+
+
+def allocate_ghb_cells(
+    allocation_option: ALLOCATION_OPTION,
+    active: GridDataArray,
+    top: GridDataArray,
+    bottom: GridDataArray,
+    head: GridDataArray,
+) -> GridDataArray:
```
Comment (luitjansl) [inline imod/prepare/topsystem/allocation.py:93]: for all these allocation functions, input arrays like "top" and "bottom" should all be xr.DataArray or xu.UgridDataArray but they should (I assume) be all of the same type. If so, could you do some validation? you can probably use the same validation function for all allocators.

**Review by luitjansl — APPROVED** [review]
looks good, some comments

File: imod/prepare/layer.py [hunk 3]
```diff
@@ -67,3 +72,14 @@ def get_lower_active_grid_cells(active: GridDataArray) -> GridDataArray:
     layer = active.coords["layer"]
     lower_active_layer = get_lower_active_layer_number(active)
     return layer == lower_active_layer
+
+
+def create_layered_top(bottom: GridDataArray, top: GridDataArray) -> GridDataArray:
+    """
+    Create a top array with layers from a single top array and a full bottom array
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/prepare/layer.py:79]: Done, extended the docstring

File: imod/prepare/topsystem/allocation.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+                active, top, bottom, bottom_elevation
+            )
+        case ALLOCATION_OPTION.first_active_to_riv_bot__drn:
+            return _allocate_cells__first_active_to_riv_bot__drn(
+                active, top, bottom, stage, bottom_elevation
+            )
+        case ALLOCATION_OPTION.at_elevation:
+            return _allocate_cells__at_elevation(top, bottom, bottom_elevation)
+        case ALLOCATION_OPTION.at_first_active:
+            return _allocate_cells__at_first_active(active)
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/prepare/topsystem/allocation.py:63]: Done

File: imod/prepare/topsystem/allocation.py (hunk 1 above)
Comment (JoerivanEngelen) [reply imod/prepare/topsystem/allocation.py:192]: bottom_elevation is used in the River package, wanted to stick to that convention as much as possible.

File: imod/prepare/topsystem/allocation.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+def allocate_rch_cells(
+    allocation_option: ALLOCATION_OPTION,
+    active: GridDataArray,
+) -> GridDataArray:
+    match allocation_option:
+        case ALLOCATION_OPTION.at_first_active:
+            return _allocate_cells__at_first_active(active)[0]
+        case _:
+            raise ValueError(
+                "Received incompatible setting for drains, only"
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/prepare/topsystem/allocation.py:117]: Good catch!

File: imod/prepare/topsystem/allocation.py (hunk 2 above)
Comment (JoerivanEngelen) [reply imod/prepare/topsystem/allocation.py:93]: Not necessarily true:

You can supply ``top`` and ``bottom`` as scalar xr.DataArray and 1D xr.DataArray with layer dimension and provide unstructured grids to the other variables.

**Pull request merged**